python/mpi_cmake_modules/documentation_builder.py

```diff
@@ -0,0 +1,529 @@
+            "   general_documentation\n\n"
+        )
+        shutil.copy(
+            resource_dir
+            / "sphinx"
+            / "sphinx"
+            / "general_documentation.rst.in",
+            str(doc_build_dir / "general_documentation.rst"),
+        )
+    return general_documentation
+
+
+def build_documentation(build_dir, project_source_dir, project_version):
+
+    #
+    # Initialize the paths
+    #
+
+    # Get the project name form the source path.
+    project_name = Path(project_source_dir).stem
+
+    # Get the build folder for the documentation.
+    doc_build_dir = Path(build_dir) / "share" / "docs" / "sphinx"
+
+    # Create the folder architecture inside the build folder.
+    shutil.rmtree(str(doc_build_dir), ignore_errors=True)
+    doc_build_dir.mkdir(parents=True, exist_ok=True)
+
+    # Get the path to resource files.
+    resource_dir = Path(_resource_path(project_source_dir))
+
+    #
+    # Parametrize the final doc layout depending we have CMake/Python/C++ api.
+    #
+
+    # String to replace in the main index.rst
+
+    cpp_api = _search_for_cpp_api(
+        doc_build_dir, project_source_dir, resource_dir
+    )
+
+    python_api = _search_for_python_api(doc_build_dir, project_source_dir)
+
+    cmake_api = _search_for_cmake_api(
+        doc_build_dir, project_source_dir, resource_dir
+    )
+
+    general_documentation = _search_for_general_documentation(
+        doc_build_dir, project_source_dir, resource_dir
+    )
+
+    #
+    # Configure the config.py and the index.rst.
+    #
+
+    # configure the index.rst.in.
+    header = "Welcome to " + project_name + "'s documentation!"
+    header += "\n" + len(header) * "=" + "\n"
+    with open(
+        str(resource_dir / "sphinx" / "sphinx" / "index.rst.in"), "rt"
+    ) as f:
+        out_text = (
+            f.read()
+            .replace("@HEADER@", header)
+            .replace("@GENERAL_DOCUMENTATION@", general_documentation)
+            .replace("@CPP_API@", cpp_api)
+            .replace("@PYTHON_API@", python_api)
+            .replace("@CMAKE_API@", cmake_api)
+        )
+    with open(str(doc_build_dir / "index.rst"), "wt") as f:
+        f.write(out_text)
+
+    # configure the config.py.in.
+    with open(
+        str(resource_dir / "sphinx" / "sphinx" / "conf.py.in"), "rt"
+    ) as f:
+        out_text = (
+            f.read()
+            .replace("@PROJECT_SOURCE_DIR@", project_source_dir)
+            .replace("@PROJECT_NAME@", project_name)
+            .replace("@PROJECT_VERSION@", project_version)
+            .replace(
+                "@DOXYGEN_XML_OUTPUT@", str(doc_build_dir / "doxygen" / "xml")
+            )
+        )
+    with open(str(doc_build_dir / "conf.py"), "wt") as f:
+        f.write(out_text)
+
+    #
+    # Copy the license and readme file.
+    #
+    readme = [
+        p.resolve()
+        for p in Path(project_source_dir).glob("*")
+        if p.suffix in [".md"] and p.name.lower() == "readme.md"
```

Comment:
The suffix check seems to be redundant here given the name check afterwards

Reply:
sure^^